8323667: Library debug files contain non-reproducible full gcc include paths

andrew-m-leonard opened the pull request:
For gcc toolchains in ALLOW_ABSOLUTE_PATHS_IN_OUTPUT=False builds, this PR finds the location of the gcc system include paths, and adds -fdebug-prefix-map flags to map them to a standard location. Thus making the debuginfo and resulting libraries reproducible when using DevKits in different path locations.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323667](https://bugs.openjdk.org/browse/JDK-8323667): Library debug files contain non-reproducible full gcc include paths (**Bug** - P4)


### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17399/head:pull/17399` \
`$ git checkout pull/17399`

Update a local copy of the PR: \
`$ git checkout pull/17399` \
`$ git pull https://git.openjdk.org/jdk.git pull/17399/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17399`

View PR using the GUI difftool: \
`$ git pr show -t 17399`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17399.diff">https://git.openjdk.org/jdk/pull/17399.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17399#issuecomment-1889389779)

erikj79 commented on make/autoconf/flags-cflags.m4:
```diff
@@ -158,6 +164,51 @@ AC_DEFUN([FLAGS_SETUP_DEBUG_SYMBOLS],
+    GCC_INCLUDE_DEBUG_MAP_FLAGS="-fdebug-prefix-map=${GCC_ROOT_INCLUDE}/=/usr/include/"
+
+    # Add gcc system include mapping => /usr/local/gcc_include
+    #   Find location of stddef.h using build C compiler
```
Why use the build compiler here?

andrew-m-leonard replied:
cos I made a misstake :-) Thanks for spotting @erikj79 
Now fixed

erikj79 replied:
Could you also update the comments that reference "build" compiler. This wording could be confusing.

andrew-m-leonard commented:
@erikj79 when you get a moment, can you re-review please? thanks

magicus commented on make/autoconf/flags-cflags.m4:
```diff
@@ -121,6 +121,12 @@ AC_DEFUN([FLAGS_SETUP_DEBUG_SYMBOLS],
       )
     fi
 
+    if test "x$DEBUG_PREFIX_CFLAGS" != x; then
```
Should this test not be inside the "if $ALLOW_ABSOLUTE_PATHS_IN_OUTPUT == false" block?

magicus replied:
The logic of this is a bit strange. I think what we really want to express is:

if ! ALLOW_ABSOLUTE_PATHS_IN_OUTPUT and -fdebug-prefix-map is a valid option for gcc then
  set -fdebug-prefix-map for both workspace_root and gcc system libraries
endif

Now this connection is kind of lost. Maybe the easiest way to do this is just to do the FLAGS_COMPILER_CHECK_ARGUMENTS separately, and reading the result of that check as a boolean variable in a separate if statement, since nesting much code inside a `IF_TRUE:` block can be hard to read. Or move the setting of the workspace_root as well to DEBUG_PREFIX_MAP_GCC_INCLUDE_PATHS, perhaps renaming it in the process, and calling it from `IF_TRUE:`.

andrew-m-leonard replied:
Hmm, i'm not sure, it makes the configure output a bit noisy:
```
checking if CC supports "-fdebug-prefix-map=/home/andrew/temurin-build/workspace/build/src/= -fdebug-prefix-map=/home/andrew/jdk21u/build/devkit/result/aarch64-linux-gnu-to-aarch64-linux-gnu/aarch64-linux-gnu/sysroot/usr/include/=/usr/include/ -fdebug-prefix-map=/home/andrew/jdk21u/build/devkit/result/aarch64-linux-gnu-to-aarch64-linux-gnu/lib/gcc/aarch64-linux-gnu/11.3.0/include/=/usr/local/gcc_include/ -fdebug-prefix-map=/home/andrew/jdk21u/build/devkit/result/aarch64-linux-gnu-to-aarch64-linux-gnu/include/c++/11.3.0/=/usr/local/gxx_include/"... yes
```
I'll try moving DEBUG_PREFIX_MAP_GCC_INCLUDE_PATHS to IF_TRUE instead

andrew-m-leonard replied:
yes, that looks better I think.
What you think?

andrew-m-leonard commented on make/autoconf/flags-cflags.m4:
```diff
@@ -158,6 +164,51 @@ AC_DEFUN([FLAGS_SETUP_DEBUG_SYMBOLS],
+
+    # Add gcc system include mapping => /usr/local/gcc_include
+    #   Find location of stddef.h using build C compiler
+    GCC_SYSTEM_INCLUDE=`$ECHO "#include <stddef.h>" | $CC $GCC_SYSROOT_PARAM -v -E - 2>&1 | $GREP stddef | $TAIL -1 | $TR -s " " | $CUT -d'"' -f2`
```
Yeah, you're right, SYSROOT_CFLAGS

magicus replied:
Actually, I meant the literal value `CFLAGS`, as in `$CC $CFLAGS`. We do not use it in the build, but autoconf uses it in the test macros it provides that we utilize, so we have to make sure it is set "good enough" to at least get the compiler to run, which at least means setting the sysroot.

andrew-m-leonard commented:
> Have you checked with `strings` if there are any other hard-coded paths, or are these really the only ones left?

I have been doing some separate machine/location builds with DevKits located in different paths and build workspace in different location, and they are now "identical" with this change.
So yes, I believe this is complete.

erikj79 reviewed (commented):
I think this looks ok but could you split the longer lines to make side by side comparisons easier in the future?

andrew-m-leonard commented:
> I think this looks ok but could you split the longer lines to make side by side comparisons easier in the future?

Yep, done, thanks Erik

andrew-m-leonard commented:
> So `$CC $CFLAGS` did not work?
@magicus 
So I purposely chose SYSROOT_CFLAGS thinking that would all that would be needed, but I think you're correct, I should use the complete CFLAGS in case a custom user extension include may have pointed at some other include path.
I have updated and tested successfully with CFLAGS_JDKLIB.
Ready for re-review, Thanks

erikj79 commented:
> > So `$CC $CFLAGS` did not work?
> 
> @magicus So I purposely chose SYSROOT_CFLAGS thinking that would all that would be needed, but I think you're correct, I should use the complete CFLAGS in case a custom user extension include may have pointed at some other include path. I have updated and tested successfully with CFLAGS_JDKLIB. Ready for re-review, Thanks

Magnus specifically meant `$CFLAGS`. It's set to `"$MACHINE_FLAG $SYSROOT_CFLAGS $USER_CFLAGS"` in `flags.m4` and meant to contain relevant flags for running checks in configure. I don't think `$CFLAGS_JDKLIB` is appropriate here even if it should work.

andrew-m-leonard commented:
> > > So `$CC $CFLAGS` did not work?
> > 
> > 
> > @magicus So I purposely chose SYSROOT_CFLAGS thinking that would all that would be needed, but I think you're correct, I should use the complete CFLAGS in case a custom user extension include may have pointed at some other include path. I have updated and tested successfully with CFLAGS_JDKLIB. Ready for re-review, Thanks
> 
> Magnus specifically meant `$CFLAGS`. It's set to `"$MACHINE_FLAG $SYSROOT_CFLAGS $USER_CFLAGS"` in `flags.m4` and meant to contain relevant flags for running checks in configure. I don't think `$CFLAGS_JDKLIB` is appropriate here even if it should work.

Ah, I thought CFLAGS was only set in the gmk files...? i'll give it a try thanks.

andrew-m-leonard commented:
> > > So `$CC $CFLAGS` did not work?
> > 
> > 
> > @magicus So I purposely chose SYSROOT_CFLAGS thinking that would all that would be needed, but I think you're correct, I should use the complete CFLAGS in case a custom user extension include may have pointed at some other include path. I have updated and tested successfully with CFLAGS_JDKLIB. Ready for re-review, Thanks
> 
> Magnus specifically meant `$CFLAGS`. It's set to `"$MACHINE_FLAG $SYSROOT_CFLAGS $USER_CFLAGS"` in `flags.m4` and meant to contain relevant flags for running checks in configure. I don't think `$CFLAGS_JDKLIB` is appropriate here even if it should work.

Thanks @erikj79 yep, that works great, thank you for pointing that out.
Updated.

erikj79 commented:
> Ah, I thought CFLAGS was only set in the gmk files...? i'll give it a try thanks.

No, we specifically only set the variable `CFLAGS` in configure because there are autoconf macros that relies on that variable. In our makefiles we avoid the standard variable `CFLAGS` and are instead choosing to be explicit with various `CFLAGS_*` and `*_CFLAGS` variables.

erikj79 reviewed (approved):
This looks good to me now.

magicus reviewed (approved):
Looks good now.

Thanks Erik for helping me clarify what I meant.

andrew-m-leonard commented:
/integrate
